src/fonduer/utils/fonduer_model.py

```diff
@@ -0,0 +1,398 @@
+    fonduer_model.candidate_extractor = CandidateExtractorUDF(
+        **model["candidate_extractor"]
+    )
+
+    # Configure logging for Fonduer
+    init_logging(log_dir="logs")
+
+    pyfunc_conf = _get_flavor_configuration(
+        model_path=model_path, flavor_name=pyfunc.FLAVOR_NAME
+    )
+    candidate_classes = fonduer_model.candidate_extractor.candidate_classes
+
+    fonduer_model.model_type = pyfunc_conf.get(MODEL_TYPE, "discriminative")
+    if fonduer_model.model_type == "discriminative":
+        emmental.init()
+        fonduer_model.featurizer = FeaturizerUDF(candidate_classes, FeatureExtractor())
+        fonduer_model.key_names = model["feature_keys"]
+        fonduer_model.word2id = model["word2id"]
+
+        # Load the disc_model
+        buffer = BytesIO()
+        buffer.write(model["disc_model"])
+        buffer.seek(0)
+        fonduer_model.disc_model = torch.load(buffer)
+    else:
+        fonduer_model.labeler = LabelerUDF(candidate_classes)
+        fonduer_model.key_names = model["labeler_keys"]
+
+        fonduer_model.lfs = model["lfs"]
+
+        fonduer_model.gen_models = []
+        for state_dict in model["gen_models_state_dict"]:
+            gen_model = LabelModel()
+            gen_model.__dict__.update(state_dict)
+            fonduer_model.gen_models.append(gen_model)
+    return _FonduerWrapper(fonduer_model)
+
+
+def log_model(
+    fonduer_model: FonduerModel,
+    artifact_path: str,
+    preprocessor: DocPreprocessor,
+    parser: Parser,
+    mention_extractor: MentionExtractor,
+    candidate_extractor: CandidateExtractor,
+    conda_env: Optional[Dict] = None,
+    code_paths: Optional[List[str]] = None,
+    model_type: Optional[str] = "discriminative",
+    labeler: Optional[Labeler] = None,
+    lfs: Optional[List[List[Callable]]] = None,
+    gen_models: Optional[List[LabelModel]] = None,
+    featurizer: Optional[Featurizer] = None,
+    disc_model: Optional[EmmentalModel] = None,
+    word2id: Optional[Dict] = None,
+) -> None:
+    """Log a Fonduer model as an MLflow artifact for the current run.
+
+    :param fonduer_model: Fonduer model to be saved.
+    :param artifact_path: Run-relative artifact path.
+    :param preprocessor: the doc preprocessor.
+    :param parser: self-explanatory
+    :param mention_extractor: self-explanatory
+    :param candidate_extractor: self-explanatory
+    :param conda_env: A dictionary representation of a Conda environment.
+    :param code_paths: A list of local filesystem paths to Python file dependencies,
+        or directories containing file dependencies. These files are prepended to the
+        system path when the model is loaded.
+    :param model_type: the model type, either "discriminative" or "generative",
+        defaults to "discriminative".
+    :param labeler: a labeler, defaults to None.
+    :param lfs: a list of list of labeling functions.
+    :param gen_models: a list of generative models, defaults to None.
```

Comment:
We use `label_model ` now?

Reply:
Good catch!
With https://github.com/HazyResearch/fonduer/pull/379, Fonduer has switched to `label_model`.

Reply:
Can we also switch from "disc model" to "end model"?

Reply:
Oh, now I can recall that I meant it.
I assume the following two scenarios.
1. a generative model as the end model
2. a discriminative model as the end model (a generative model is used purely as a label model)

In either case, the end model should be packaged, which is the generative model in the scenario 1 and the discriminative model in the scenario 2.

Snorkel changed the terminology (from gen_model vs. disc_model to label_model vs. end_model) to make them more clearer (what the models are actually intended to do vs. how they do it.)
From Snorkel perspective, this makes sense. However, from Fonduer (or an application that uses Snorkel) perspective, it may not if you use a generative model as the end model.

Having said, let's switch to label_model vs. end_model to avoid any confusion.

Reply:
label_model vs. emmental_model sounds clearer in Fonduer.

Reply:
Renamed `gen_model` and `disc_model`, to `label_model` and `emmental_model`, respectively.